iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -0,0 +1,208 @@
+        return false;
+      }
+
+      // We want to check the operand _types_, but don't care if the actual
+      // operand references differ (as they live in separate modules anyway).
+      if (!std::equal(lhsOp.operand_type_begin(), lhsOp.operand_type_end(),
+                      rhsOp.operand_type_begin())) {
+        return false;
+      }
+
+      // If the operations have regions, recurse into them (depth-first).
+      if (lhsOp.getNumRegions() != rhsOp.getNumRegions()) {
+        return false;
+      }
+      auto lhsRegions = lhsOp.getRegions();
+      auto rhsRegions = rhsOp.getRegions();
+      for (int i = 0; i < lhsRegions.size(); ++i) {
+        if (!areRegionsEquivalent(&lhsRegions[i], &rhsRegions[i])) {
+          return false;
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+bool areExecutablesEquivalent(ExecutableOp lhs, ExecutableOp rhs) {
+  auto lhsModule = lhs.getInnerModule();
+  auto rhsModule = rhs.getInnerModule();
+
+  // Must have the same number of entry point ops, with the same attributes.
+  // Entry point op symbol names are expected to differ, that won't affect
+  // equivalence.
+  auto lhsEntryOps = llvm::to_vector<1>(lhsModule.getOps<DispatchEntryOp>());
+  auto rhsEntryOps = llvm::to_vector<1>(rhsModule.getOps<DispatchEntryOp>());
+  if (lhsEntryOps.size() != rhsEntryOps.size()) {
+    return false;
+  }
+  for (int i = 0; i < lhsEntryOps.size(); ++i) {
+    if (lhsEntryOps[i].getAttrs() != rhsEntryOps[i].getAttrs()) {
+      return false;
+    }
+  }
+
+  // Must have the same number of functions, with each listed in the same order
+  // and with equivalent regions inside.
+  auto lhsFuncOps = llvm::to_vector<1>(lhsModule.getOps<FuncOp>());
+  auto rhsFuncOps = llvm::to_vector<1>(rhsModule.getOps<FuncOp>());
+  if (lhsFuncOps.size() != rhsFuncOps.size()) {
+    return false;
+  }
+  for (int i = 0; i < lhsFuncOps.size(); ++i) {
+    auto lhsRegion = lhsFuncOps[i].getCallableRegion();
+    auto rhsRegion = rhsFuncOps[i].getCallableRegion();
+    return areRegionsEquivalent(lhsRegion, rhsRegion);
+  }
+
+  return true;
+}
+
+}  // namespace
+
+class DeduplicateExecutablesPass
+    : public PassWrapper<DeduplicateExecutablesPass, OperationPass<ModuleOp>> {
+ public:
+  explicit DeduplicateExecutablesPass() {}
+  DeduplicateExecutablesPass(const DeduplicateExecutablesPass &pass) {}
+
+  void runOnOperation() override {
+    auto moduleOp = getOperation();
+    auto executableOps = llvm::to_vector<8>(moduleOp.getOps<ExecutableOp>());
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
+
+    SmallVector<ExecutableOp, 3> duplicateExecutableOps;
+    DenseMap<Attribute, Attribute> entryPointRefReplacements;
+
+    // For each executable, find the first executable which it is equivalent to.
+    for (int i = executableOps.size() - 1; i >= 0; --i) {
+      auto executableOp = executableOps[i];
+
+      for (int j = 0; j < i; ++j) {
+        auto comparisonExecutableOp = executableOps[j];
```

Comment:
you can hint to the reader what this algorithm is doing by picking more relevant names for these; like `executableOp` -> `duplicateExecutableOp` and `comparisonExecutableOp` -> `referenceExecutableOp` indicating that the dupe is being checked against each possible reference and if a match is found then the duplicate will be replaced with the reference.